### PR TITLE
performance improvements

### PR DESCRIPTION
**Pull request by PatrickvonDeetzen**
TL;DR
- create json with special characters is now significantly faster (changes made in function "escapeandencode")
- updated performance test & added new test sample
 
===========

Overall summary:
I have used the included performance test (with largely increased sample size) to get an approximate idea for the performance change. Without special characters in json performance does not change. With at least one special character in json the improvements can be measured (depending on overall json size). When "/" is the only special character in json then it speeds up significantly, because "/" was already no longer threated as special character in funtion "escapeandencode".

Changes in json create:
+33% more op/sec - flicker
+53% more op/sec - Google Maps Basic Query
+27% more op/sec - Youtube
+21% more op/sec - facebook
+44% more op/sec - Tweeter 1
+23% more op/sec - Tweeter 2
+53% more op/sec - Tweeter 3
+18% more op/sec - Tweeter 4
+17% more op/sec - Sample Konfabulator Widget
+36% more op/sec - cofaxCDS
+32% more op/sec - Cursor Rows
+31% more op/sec - Amazon S3
+24% more op/sec - dropBox
+32% more op/sec - tropical index
+41% more op/sec - jSON Benchmark
+15% more op/sec - Map Quest
+12% more op/sec - Bing
+28% more op/sec - JSONAPI
+36% more op/sec - Child-Care Informations
=> other samples have no significant change in op/sec (less than +5%).

=====================

More detailed explanation:

*nfjsoncreate.prg / variable "specialChars"*
- removed multiple instances of the same character from variable "specialChars"
- removed character "/" from variable "specialChars", because for this character there is already no more encoding done in function "escapeandencode". Depending on the data some strings do not call the function "escapeandencode" anymore. 

*nfjsoncreate.prg / function "escapeandencode"*
- removed unnecessary checks for "mustencode", because none of the special characters will be removed till then (even if it would, then the call of "RETURN" would throw an error, because a string is expected as return value and not a boolean ".F." ...)
- on modern processors (at least from the 2000s) can handle far more then one string modification per cpu cycle. On top of that using through FOR ... ENDFOR and having to incrementally step from x = 0 to 31 takes additional 32 cpu cycles that slow us down even more.
- to still preserve some kind of readability all the STRTRAN() calls to manipulate the string are combined in 4 big inline executions. By tracking this function with "SET COVERAGE TO" this new code was executed about 8x faster than the original code.

*nfjsonperftest.prg*
- increased the default sample size from 10 to 75, because this gives us more accurate execution times for read & create. When i started testing some high & low execution times were off by 2-5 times. no idea how i should compare these results.
- for better results i needed even larger sample sizes, but everytime i changed the value in "nfjsonperftest.prg" i had to close VFP, delete the FXP and start it up again so the new values would be used. this pissed me off, so now you can see and change the sample size interactively in the form. 
- since i could now test with bigger sample sizes the results got far more consistent  and meaningful. Because larger sample sizes take longer to execute i needed some kind of indicator to see if a single execution is finished. So now the read & create informations are shown in red while execution is running.
- added one more json sample with special characters used in most of the strings.
- changed the size of the form, because it bothered me to scroll through the ~20 samples on my 27" display.

==============

Feel free to comment these changes or roast me (if necessary). I hope you all can somewhat understand my english, since i try my best to be clear.

P.S.: Lots of kudos to all the bigbrain creators and maintainers that made nfjson to what it is today. I know i have already learned a lot by digging through some parts of this code. ;-)

### CODE REVIEW

**Review by nfoxdev — CHANGES_REQUESTED** [review]
Hello Patrick, your changes look great - though I have not confirmed the speed gains yet-. I just did a check using tests\escapetest.prg  feeding a string with non printable characters ( 
addproperty(pair , [test], [\\"""he\"\\]+chr(13)+chr(20)+chr(22)+["""\///""\"\\"/"ll/\"o/"\"\\\\""\"] )  but it fails. As soon as you can fix we can do the merge!

**Comment by PatrickvonDeetzen** [comment]
Hi Marco, i assume "it fails" means it does not the expected outcome, because at least on my machine i didnt get an error message. 

But i could see that every special character was encoded to "\u0000" which is definetly wrong.   Too much late night copy&paste, my bad. After rechecking all special characters again i just commited a fix for that problem.

**Review by nfoxdev — COMMENTED** [review]
Patrick your changes were merged. Thanks for your collaboration.!
